bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/WrappedScheduledExecutorService.java

```diff
@@ -0,0 +1,60 @@
+/**
+ * Copyright (c) 2014,2017 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.core.common;
+
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.ThreadFactory;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class wraps the ScheduledThreadPoolExecutor to implement the {@link #afterExecute(Runnable, Throwable)} method
+ * and log the error. The error will otherwise go unnoticed because on an exception thrown in the thread it will simply
+ * end with no logging or whatsoever unless the user handles it. This wrappers removes the burden by the user to always
+ * catch errors in scheduled threads for logging and it also catches unhandled exceptions that can be the cause of very
+ * hard to catch bugs because no error is ever shown.
+ * The implementation follows the implementation suggestion as given in the JavaDoc of the method.
+ *
+ * @author Hilbrand Bouwkamp - Initial contribution
+ */
+class WrappedScheduledExecutorService extends ScheduledThreadPoolExecutor {
+
+    final static Logger logger = LoggerFactory.getLogger(WrappedScheduledExecutorService.class);
+
+    public WrappedScheduledExecutorService(int corePoolSize, ThreadFactory threadFactory) {
+        super(corePoolSize, threadFactory);
+    }
+
+    @Override
+    protected void afterExecute(Runnable r, Throwable t) {
+        Throwable actualThrowable = t;
+        if (actualThrowable == null && r instanceof Future<?>) {
+            try {
+                ((Future<?>) r).get();
+            } catch (CancellationException ce) {
+                actualThrowable = ce;
+            } catch (ExecutionException ee) {
+                actualThrowable = ee.getCause();
```

Comment:
What has been the reason to use the cause only? Shouldn't we keep the execution exception in the stack trace? Didn't we see "has caused by" and the cause exception in the log?

Reply:
The ExecutionException is thrown by the get method when the thread method was stopped on an exception. The cause is that exception. For logging I don't think it adds value to also keep te part of the ExecutionException as it doesn't given extra information. It's just a way the get method uses to pass the exception.
Related to your other comment regarding the implementation suggestion about the line to remove. The JavaDoc of the method contains the complete implementation and I just used that implementation suggestion. That's the reason I explicitly added the line (although it might not add value, so I removed it). But that implementation suggestion also uses the getCause(), hence I used it here too.

Reply:
The question is as we are using
```
logger.error("Scheduled thread ended with an exception: ", actualThrowable);
```

does need the developer that wants to fix a bug needs to know if `actualThrowable` is the one that has been received by the function argument (`t`) or is the one that has been the cause of the `ExecutionException` of a future?

Does he "see" any difference?

Reply:
Actually now that I think of it the log message should probably not log to `error` but to `warn`.